src/FlowDiagnostics.jl

```diff
@@ -40,15 +40,64 @@ function add_background_fields(model)
+#+++ Richardson number
+@inline ψ²(i, j, k, grid, ψ) = @inbounds ψ[i, j, k]^2
+
+"""
+Get `w` from `û`, `v̂`, `ŵ` and based on the direction given by the unit vector `(k_x, k_y, k_z)`.
+"""
+@inline function w²_from_u⃗_tilted_ccc(i, j, k, grid, û, v̂, ŵ, k_x, k_y, k_z)
+    û = ℑxᶜᵃᵃ(i, j, k, grid, û) # F, C, C  → C, C, C
+    v̂ = ℑyᵃᶜᵃ(i, j, k, grid, v̂) # C, F, C  → C, C, C
+    ŵ = ℑzᵃᵃᶜ(i, j, k, grid, ŵ) # C, C, F  → C, C, C
+    return (û * k_x + v̂ * k_y + ŵ * k_z)^2
+end
+
+"""
+Return the (true) horizontal velocity magnitude.
+"""
+@inline function uₕ_norm_ccc(i, j, k, grid, û, v̂, ŵ, k_x, k_y, k_z)
+    û² = ℑxᶜᵃᵃ(i, j, k, grid, ψ², û) # F, C, C  → C, C, C
+    v̂² = ℑyᵃᶜᵃ(i, j, k, grid, ψ², v̂) # C, F, C  → C, C, C
+    ŵ² = ℑzᵃᵃᶜ(i, j, k, grid, ψ², ŵ) # C, C, F  → C, C, C
+    return √(û² + v̂² + ŵ² - w²_from_u⃗_tilted_ccc(i, j, k, grid, û, v̂, ŵ, k_x, k_y, k_z))
+end
 
-    dBdz_tot = ∂z(b) + N²_bg
-    dUdz_tot = ∂z(u) + dUdz_bg
-    dVdz_tot = ∂z(v) + dVdz_bg
+@inline function richardson_number_ccf(i, j, k, grid, û, v̂, ŵ, b, params)
 
-    return dBdz_tot / (dUdz_tot^2 + dVdz_tot^2)
+    dbdx̂ = ℑxzᶜᵃᶠ(i, j, k, grid, ∂xᶠᶜᶜ, b) # C, C, C  → F, C, C → C, C, F
+    dbdŷ = ℑyzᵃᶜᶠ(i, j, k, grid, ∂yᶜᶠᶜ, b) # C, C, C  → C, F, C → C, C, F
+    dbdẑ = ∂zᶜᶜᶠ(i, j, k, grid, b) # C, C, C  → C, C, F
+    dbdz = dbdx̂ * params.vertical_dir_x + dbdŷ * params.vertical_dir_y + dbdẑ * params.vertical_dir_z
+
+    duₕdx̂ = ℑxᶜᵃᵃ(i, j, k, grid, ∂xᶠᶜᶜ, uₕ_norm_ccc, û, v̂, ŵ, params.vertical_dir_x, params.vertical_dir_y, params.vertical_dir_z)
+    duₕdŷ = ℑyᵃᶜᵃ(i, j, k, grid, ∂yᶜᶠᶜ, uₕ_norm_ccc, û, v̂, ŵ, params.vertical_dir_x, params.vertical_dir_y, params.vertical_dir_z)
+    duₕdẑ = ∂zᶜᶜᶠ(i, j, k, grid, uₕ_norm_ccc, û, v̂, ŵ, params.vertical_dir_x, params.vertical_dir_y, params.vertical_dir_z)
+    duₕdz = duₕdx̂ * params.vertical_dir_x + duₕdŷ * params.vertical_dir_y + duₕdẑ * params.vertical_dir_z
+
+    return dbdz / duₕdz^2
 end
+function RichardsonNumber(model; location = (Center, Center, Face), add_background=true)
+    validate_location(location, "RichardsonNumber", (Center, Center, Face))
+
+    if (model isa NonhydrostaticModel) & add_background
+        full_fields = add_background_fields(model)
+        u, v, w, b = full_fields.u, full_fields.v, full_fields.w, full_fields.b
+    else
+        u, v, w = model.velocities
+        b = model.tracers.b
+    end
+
+    true_vertical_direction =  model.buoyancy.gravity_unit_vector
+    if true_vertical_direction isa Oceananigans.Grids.ZDirection
+        vertical_dir_x, vertical_dir_y, vertical_dir_z = (0, 0, 1)
+    else
+        vertical_dir_x, vertical_dir_y, vertical_dir_z = true_vertical_direction
+    end
+    return KernelFunctionOperation{Center, Center, Face}(richardson_number_ccf, model.grid;
+                                                     computed_dependencies=(u, v, w, b), parameters=(; vertical_dir_x, vertical_dir_y, vertical_dir_z))
```

Comment:
why not just pass `gravity_unit_vector` in? It's more general, you won't have to write new if statements for every new unit vector type.

Reply:
As in substitute `parameters=(; vertical_dir_x, vertical_dir_y, vertical_dir_z)` with `parameters=gravity_unity_vector`?

If so, when I tried that on the GPU I got errors, so I thought it wasn't possible. But maybe I was doing something wrong?

Reply:
It should be possible. For example, the tendency kernels in Oceananigans also use `gravity_unit_vector`.

Reply:
Hmm... I may have done something wrong when I tried this then. I'll try again since I agree it would be better.

Thanks for the tip.

Reply:
Done!